In [195]:
import numpy as np
import pandas as pd
import matplotlib as plt

data = pd.read_csv('family_data.csv')

In [196]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 12 columns):
family_id    5000 non-null int64
choice_0     5000 non-null int64
choice_1     5000 non-null int64
choice_2     5000 non-null int64
choice_3     5000 non-null int64
choice_4     5000 non-null int64
choice_5     5000 non-null int64
choice_6     5000 non-null int64
choice_7     5000 non-null int64
choice_8     5000 non-null int64
choice_9     5000 non-null int64
n_people     5000 non-null int64
dtypes: int64(12)
memory usage: 468.8 KB


In [197]:
total_peeps = data.n_people.sum()
total_peeps

21003

\begin{equation*}
Loss=\sum_{f=1}^F pref_f(n_f) + \sum_{d=1}^D (\alpha n_d^2 - \beta n_d) e^(\frac{1}{2} + \frac{ \left\lvert{n_d-n_{d+1}}\right\rvert}{50 } )
\end{equation*}

Where F=5000 (number of families), D=100 (number of days), and n is the number of people in a family or on a given day (based on the subscript).

We begin by minimizing the right-most term of the loss function. This is done by having as many days as possible where $\alpha n_d^2 - \beta n_d = 0$ and then by keeping the exponential term $\frac{ \left\lvert{n_d-n_{d+1}}\right\rvert}{50 }$ equal to zero on all remaining days where $n_d > 125$.

In [198]:
def acct_cost(sced):
    
    return sum( [ ((sced[i]-125)/400)*sced[i]**(0.5+(abs(sced[i]-sced[i+1])/50)) for i in range(len(sced)-1)])+((sced[99]-125)/400)*sced[len(sced)-1]**(0.5)


In [199]:
best_sced=np.ones(50)*125
best_sced=np.append(best_sced, np.ones(49)*295)
best_sced=np.append(best_sced, np.ones(1)*298)

In [200]:
acct_cost(best_sced)

368.115694310885

The minimum accounting cost is \$368.12. What is the associated preference cost?

In [201]:
data.head()

family_id  choice_0  choice_1  choice_2  choice_3  choice_4  choice_5  \
0          0        52        38        12        82        33        75   
1          1        26         4        82         5        11        47   
2          2       100        54        25        12        27        82   
3          3         2        95         1        96        32         6   
4          4        53         1        47        93        26         3   

   choice_6  choice_7  choice_8  choice_9  n_people  
0        64        76        10        28         4  
1        38         6        66        61         4  
2        10        89        80        33         3  
3        40        31         9        59         2  
4        46        16        42        39         4

In [202]:
def family_cost(choice, n):
    pref={
        0:0,
        1:50,
        2:50+9*n,
        3:100+9*n,
        4:200+9*n,
        5:200+18*n,
        6:300+18*n,
        7:300+36*n,
        8:400+36*n,
        9:500+235*n,
        10:500+434*n
    }
    return pref.get(choice)

In [203]:
data=data.assign(day=data.choice_0)
data=data.assign(pref_rank=0)

In [216]:
sced=[data[data.day==j].n_people.sum() for j in range(1,101)]

In [205]:
def balance(day,sced, round):
    openings=best_sced-sced
    diff=-openings[day-1]
    
    if diff > 1:
        
        fam=data[(data.day==day)&(data.n_people<=diff)]
        k=0
        while openings[fam.loc[fam.index[k]][round+1]-1]<2:
            k=k+1
            print("k= "+str(k)+", "+str(len(fam)))
            if k==len(fam):
                return print("Day "+str(day)+"is almost balanced")
        data.loc[fam.loc[fam.index[k]].family_id].day=fam.loc[fam.index[k]][round+1]
        data.loc[fam.loc[fam.index[k]].family_id].pref_rank=round
        sced[day-1]=data[data.day==day].n_people.sum()
        sced[fam.loc[fam.index[k]][round+1]-1]=data[data.day==fam.loc[fam.index[k]][round+1]].n_people.sum()
        return balance(day,sced,round )
    return print("Day "+str(day)+"is balanced")

In [209]:
for m in range(1,10):
    [balance(day, sced, m) for day in range(1,101)]

k= 1, 129
k= 2, 129
k= 3, 129
k= 4, 129
k= 5, 129
k= 6, 129
k= 7, 129
k= 8, 129
k= 9, 129
k= 10, 129
k= 11, 129
k= 12, 129
k= 13, 129
k= 14, 129
k= 15, 129
k= 16, 129
k= 17, 129
k= 18, 129
k= 19, 129
k= 20, 129
k= 21, 129
k= 22, 129
k= 23, 129
k= 24, 129
k= 25, 129
k= 26, 129
k= 27, 129
k= 28, 129
k= 29, 129
k= 30, 129
k= 31, 129
k= 32, 129
k= 33, 129
k= 34, 129
k= 35, 129
k= 36, 129
k= 37, 129
k= 38, 129
k= 39, 129
k= 40, 129
k= 41, 129
k= 42, 129
k= 43, 129
k= 44, 129
k= 45, 129
k= 46, 129
k= 47, 129
k= 48, 129
k= 49, 129
k= 50, 129
k= 51, 129
k= 52, 129
k= 53, 129
k= 54, 129
k= 55, 129
k= 56, 129
k= 57, 129
k= 58, 129
k= 59, 129
k= 60, 129
k= 61, 129
k= 62, 129
k= 63, 129
k= 64, 129
k= 65, 129
k= 66, 129
k= 67, 129
k= 68, 129
k= 69, 129
k= 70, 129
k= 71, 129
k= 72, 129
k= 73, 129
k= 74, 129
k= 75, 129
k= 76, 129
k= 77, 129
k= 78, 129
k= 79, 129
k= 80, 129
k= 81, 129
k= 82, 129
k= 83, 129
k= 84, 129
k= 85, 129
k= 86, 129
k= 87, 129
k= 88, 129
k= 89, 129
k= 90, 129
k= 91, 129
k= 92, 1

k= 1, 38
k= 2, 38
k= 3, 38
k= 4, 38
k= 5, 38
k= 6, 38
k= 7, 38
k= 8, 38
k= 9, 38
k= 10, 38
k= 11, 38
k= 12, 38
k= 13, 38
k= 14, 38
k= 15, 38
k= 16, 38
k= 17, 38
k= 18, 38
k= 19, 38
k= 20, 38
k= 21, 38
k= 22, 38
k= 23, 38
k= 24, 38
k= 25, 38
k= 26, 38
k= 27, 38
k= 28, 38
k= 29, 38
k= 30, 38
k= 31, 38
k= 32, 38
k= 33, 38
k= 34, 38
k= 35, 38
k= 36, 38
k= 37, 38
k= 38, 38
Day 46is almost balanced
k= 1, 36
k= 2, 36
k= 3, 36
k= 4, 36
k= 5, 36
k= 6, 36
k= 7, 36
k= 8, 36
k= 9, 36
k= 10, 36
k= 11, 36
k= 12, 36
k= 13, 36
k= 14, 36
k= 15, 36
k= 16, 36
k= 17, 36
k= 18, 36
k= 19, 36
k= 20, 36
k= 21, 36
k= 22, 36
k= 23, 36
k= 24, 36
k= 25, 36
k= 26, 36
k= 27, 36
k= 28, 36
k= 29, 36
k= 30, 36
k= 31, 36
k= 32, 36
k= 33, 36
k= 34, 36
k= 35, 36
k= 36, 36
Day 47is almost balanced
Day 48is balanced
Day 49is balanced
Day 50is balanced
Day 51is balanced
Day 52is balanced
Day 53is balanced
Day 54is balanced
Day 55is balanced
Day 56is balanced
Day 57is balanced
Day 58is balanced
Day 59is balanced
Day 60is bal

k= 85, 129
k= 86, 129
k= 87, 129
k= 88, 129
k= 89, 129
k= 90, 129
k= 91, 129
k= 92, 129
k= 93, 129
k= 94, 129
k= 95, 129
k= 96, 129
k= 97, 129
k= 98, 129
k= 99, 129
k= 100, 129
k= 101, 129
k= 102, 129
k= 103, 129
k= 104, 129
k= 105, 129
k= 106, 129
k= 107, 129
k= 108, 129
k= 109, 129
k= 110, 129
k= 111, 129
k= 112, 129
k= 113, 129
k= 114, 129
k= 115, 129
k= 116, 129
k= 117, 129
k= 118, 129
k= 119, 129
k= 120, 129
k= 121, 129
k= 122, 129
k= 123, 129
k= 124, 129
k= 125, 129
k= 126, 129
k= 127, 129
k= 128, 129
k= 129, 129
Day 1is almost balanced
Day 2is balanced
k= 1, 44
k= 2, 44
k= 3, 44
k= 4, 44
k= 5, 44
k= 6, 44
k= 7, 44
k= 8, 44
k= 9, 44
k= 10, 44
k= 11, 44
k= 12, 44
k= 13, 44
k= 14, 44
k= 15, 44
k= 16, 44
k= 17, 44
k= 18, 44
k= 19, 44
k= 20, 44
k= 21, 44
k= 22, 44
k= 23, 44
k= 24, 44
k= 25, 44
k= 26, 44
k= 27, 44
k= 28, 44
k= 29, 44
k= 30, 44
k= 31, 44
k= 32, 44
k= 33, 44
k= 34, 44
k= 35, 44
k= 36, 44
k= 37, 44
k= 38, 44
k= 39, 44
k= 40, 44
k= 41, 44
k= 42, 44
k= 43, 44
k= 44, 44
Day

k= 18, 47
k= 19, 47
k= 20, 47
k= 21, 47
k= 22, 47
k= 23, 47
k= 24, 47
k= 25, 47
k= 26, 47
k= 27, 47
k= 28, 47
k= 29, 47
k= 30, 47
k= 31, 47
k= 32, 47
k= 33, 47
k= 34, 47
k= 35, 47
k= 36, 47
k= 37, 47
k= 38, 47
k= 39, 47
k= 40, 47
k= 41, 47
k= 42, 47
k= 43, 47
k= 44, 47
k= 45, 47
k= 46, 47
k= 47, 47
Day 17is almost balanced
k= 1, 40
k= 2, 40
k= 3, 40
k= 4, 40
k= 5, 40
k= 6, 40
k= 7, 40
k= 8, 40
k= 9, 40
k= 10, 40
k= 11, 40
k= 12, 40
k= 13, 40
k= 14, 40
k= 15, 40
k= 16, 40
k= 17, 40
k= 18, 40
k= 19, 40
k= 20, 40
k= 21, 40
k= 22, 40
k= 23, 40
k= 24, 40
k= 25, 40
k= 26, 40
k= 27, 40
k= 28, 40
k= 29, 40
k= 30, 40
k= 31, 40
k= 32, 40
k= 33, 40
k= 34, 40
k= 35, 40
k= 36, 40
k= 37, 40
k= 38, 40
k= 39, 40
k= 40, 40
Day 18is almost balanced
k= 1, 45
k= 2, 45
k= 3, 45
k= 4, 45
k= 5, 45
k= 6, 45
k= 7, 45
k= 8, 45
k= 9, 45
k= 10, 45
k= 11, 45
k= 12, 45
k= 13, 45
k= 14, 45
k= 15, 45
k= 16, 45
k= 17, 45
k= 18, 45
k= 19, 45
k= 20, 45
k= 21, 45
k= 22, 45
k= 23, 45
k= 24, 45
k= 25, 45
k= 26, 45
k= 27, 4

k= 99, 129
k= 100, 129
k= 101, 129
k= 102, 129
k= 103, 129
k= 104, 129
k= 105, 129
k= 106, 129
k= 107, 129
k= 108, 129
k= 109, 129
k= 110, 129
k= 111, 129
k= 112, 129
k= 113, 129
k= 114, 129
k= 115, 129
k= 116, 129
k= 117, 129
k= 118, 129
k= 119, 129
k= 120, 129
k= 121, 129
k= 122, 129
k= 123, 129
k= 124, 129
k= 125, 129
k= 126, 129
k= 127, 129
k= 128, 129
k= 129, 129
Day 1is almost balanced
Day 2is balanced
k= 1, 44
k= 2, 44
k= 3, 44
k= 4, 44
k= 5, 44
k= 6, 44
k= 7, 44
k= 8, 44
k= 9, 44
k= 10, 44
k= 11, 44
k= 12, 44
k= 13, 44
k= 14, 44
k= 15, 44
k= 16, 44
k= 17, 44
k= 18, 44
k= 19, 44
k= 20, 44
k= 21, 44
k= 22, 44
k= 23, 44
k= 24, 44
k= 25, 44
k= 26, 44
k= 27, 44
k= 28, 44
k= 29, 44
k= 30, 44
k= 31, 44
k= 32, 44
k= 33, 44
k= 34, 44
k= 35, 44
k= 36, 44
k= 37, 44
k= 38, 44
k= 39, 44
k= 40, 44
k= 41, 44
k= 42, 44
k= 43, 44
k= 44, 44
Day 3is almost balanced
Day 4is balanced
k= 1, 4
k= 2, 4
k= 3, 4
k= 4, 4
Day 5is almost balanced
Day 6is balanced
Day 7is balanced
Day 8is balanced
Day 9is b

k= 24, 46
k= 25, 46
k= 26, 46
k= 27, 46
k= 28, 46
k= 29, 46
k= 30, 46
k= 31, 46
k= 32, 46
k= 33, 46
k= 34, 46
k= 35, 46
k= 36, 46
k= 37, 46
k= 38, 46
k= 39, 46
k= 40, 46
k= 41, 46
k= 42, 46
k= 43, 46
k= 44, 46
k= 45, 46
k= 46, 46
Day 12is almost balanced
Day 13is balanced
Day 14is balanced
k= 1, 4
k= 2, 4
k= 3, 4
k= 4, 4
Day 15is almost balanced
Day 16is balanced
k= 1, 47
k= 2, 47
k= 3, 47
k= 4, 47
k= 5, 47
k= 6, 47
k= 7, 47
k= 8, 47
k= 9, 47
k= 10, 47
k= 11, 47
k= 12, 47
k= 13, 47
k= 14, 47
k= 15, 47
k= 16, 47
k= 17, 47
k= 18, 47
k= 19, 47
k= 20, 47
k= 21, 47
k= 22, 47
k= 23, 47
k= 24, 47
k= 25, 47
k= 26, 47
k= 27, 47
k= 28, 47
k= 29, 47
k= 30, 47
k= 31, 47
k= 32, 47
k= 33, 47
k= 34, 47
k= 35, 47
k= 36, 47
k= 37, 47
k= 38, 47
k= 39, 47
k= 40, 47
k= 41, 47
k= 42, 47
k= 43, 47
k= 44, 47
k= 45, 47
k= 46, 47
k= 47, 47
Day 17is almost balanced
k= 1, 40
k= 2, 40
k= 3, 40
k= 4, 40
k= 5, 40
k= 6, 40
k= 7, 40
k= 8, 40
k= 9, 40
k= 10, 40
k= 11, 40
k= 12, 40
k= 13, 40
k= 14, 40
k= 15, 40
k= 16, 

k= 48, 48
Day 33is almost balanced
Day 34is balanced
Day 35is balanced
Day 36is balanced
Day 37is balanced
k= 1, 35
k= 2, 35
k= 3, 35
k= 4, 35
k= 5, 35
k= 6, 35
k= 7, 35
k= 8, 35
k= 9, 35
k= 10, 35
k= 11, 35
k= 12, 35
k= 13, 35
k= 14, 35
k= 15, 35
k= 16, 35
k= 17, 35
k= 18, 35
k= 19, 35
k= 20, 35
k= 21, 35
k= 22, 35
k= 23, 35
k= 24, 35
k= 25, 35
k= 26, 35
k= 27, 35
k= 28, 35
k= 29, 35
k= 30, 35
k= 31, 35
k= 32, 35
k= 33, 35
k= 34, 35
k= 35, 35
Day 38is almost balanced
k= 1, 38
k= 2, 38
k= 3, 38
k= 4, 38
k= 5, 38
k= 6, 38
k= 7, 38
k= 8, 38
k= 9, 38
k= 10, 38
k= 11, 38
k= 12, 38
k= 13, 38
k= 14, 38
k= 15, 38
k= 16, 38
k= 17, 38
k= 18, 38
k= 19, 38
k= 20, 38
k= 21, 38
k= 22, 38
k= 23, 38
k= 24, 38
k= 25, 38
k= 26, 38
k= 27, 38
k= 28, 38
k= 29, 38
k= 30, 38
k= 31, 38
k= 32, 38
k= 33, 38
k= 34, 38
k= 35, 38
k= 36, 38
k= 37, 38
k= 38, 38
Day 39is almost balanced
k= 1, 38
k= 2, 38
k= 3, 38
k= 4, 38
k= 5, 38
k= 6, 38
k= 7, 38
k= 8, 38
k= 9, 38
k= 10, 38
k= 11, 38
k= 12, 38
k= 13, 38
k= 14, 38


k= 1, 129
k= 2, 129
k= 3, 129
k= 4, 129
k= 5, 129
k= 6, 129
k= 7, 129
k= 8, 129
k= 9, 129
k= 10, 129
k= 11, 129
k= 12, 129
k= 13, 129
k= 14, 129
k= 15, 129
k= 16, 129
k= 17, 129
k= 18, 129
k= 19, 129
k= 20, 129
k= 21, 129
k= 22, 129
k= 23, 129
k= 24, 129
k= 25, 129
k= 26, 129
k= 27, 129
k= 28, 129
k= 29, 129
k= 30, 129
k= 31, 129
k= 32, 129
k= 33, 129
k= 34, 129
k= 35, 129
k= 36, 129
k= 37, 129
k= 38, 129
k= 39, 129
k= 40, 129
k= 41, 129
k= 42, 129
k= 43, 129
k= 44, 129
k= 45, 129
k= 46, 129
k= 47, 129
k= 48, 129
k= 49, 129
k= 50, 129
k= 51, 129
k= 52, 129
k= 53, 129
k= 54, 129
k= 55, 129
k= 56, 129
k= 57, 129
k= 58, 129
k= 59, 129
k= 60, 129
k= 61, 129
k= 62, 129
k= 63, 129
k= 64, 129
k= 65, 129
k= 66, 129
k= 67, 129
k= 68, 129
k= 69, 129
k= 70, 129
k= 71, 129
k= 72, 129
k= 73, 129
k= 74, 129
k= 75, 129
k= 76, 129
k= 77, 129
k= 78, 129
k= 79, 129
k= 80, 129
k= 81, 129
k= 82, 129
k= 83, 129
k= 84, 129
k= 85, 129
k= 86, 129
k= 87, 129
k= 88, 129
k= 89, 129
k= 90, 129
k= 91, 129
k= 92, 1

In [211]:
def overcap(day,sced):
    openings=best_sced-sced
    if openings[day-1]<-2:
        new_day=np.where(openings==np.max(openings))[0][0]
        ind=data[(data.day==day)&(data.n_people<=-openings[day-1])].index[0]
        data.loc[ind].day=new_day+1
        data.loc[ind].rank=10
        sced[day-1]=data[data.day==day].n_people.sum()
        sced[new_day]=data[data.day==new_day+1].n_people.sum()
        return overcap(day,sced)
        

In [212]:
%%capture
[overcap(h,sced) for h in range(1,101)]

In [213]:
data[(data.day==50) & (data.choice_0!=50)]

family_id  choice_0  choice_1  choice_2  choice_3  choice_4  choice_5  \
733         733         1        50        60        48        26        45   
1595       1595         1        50        49        12        34        81   
1597       1597         1        50        97        96        32         9   
3307       3307         1        50        94        90        44        88   
3871       3871         2        50        61         7        66        11   
4458       4458         5        50        32        45        46         1   
4696       4696         1        50         8        74         7         4   
4814       4814         1        50        42        70        29        47   

      choice_6  choice_7  choice_8  choice_9  n_people  day  pref_rank  
733         74        29        17        42         4   50          1  
1595        57        28         8        36         2   50          1  
1597         4        47         2        26         8   50          1  
3307         2        22        60         4         3   50          1  
3871        39        75        96        70         3   50          1  
4458        67        14        86        74         5   50          1  
4696        67        81        30        15         6   50          1  
4814         9        88         7        60         3   50          1

In [214]:
s=100
data[(data.day==s) & (data.choice_0!=s)]

family_id  choice_0  choice_1  choice_2  choice_3  choice_4  choice_5  \
70           70        32         1        11         5         7       100   
102         102        31        29        13        46        89        24   
129         129        17        43        32        61        46       100   
196         196        26       100        60         7        54        61   
224         224         4        66        40       100        87        12   
282         282        17        12        26        89         4        54   
309         309        19         4        31        96        66        18   
464         464        46         1        43        96        19        68   
531         531        45        61        75         2        95        57   
558         558        18        20         3        94        41        58   
638         638        26        16         6        46         1        23   
778         778         3        18        17        32        53        34   
798         798        45        35        25        20         3         4   
814         814        27        59        47        80        26        32   
888         888        25        30         1        16        42        39   
1011       1011        24         1        50        38        87        27   
1182       1182        11        16        18        89        80         1   
1214       1214         3       100         1        95        19        68   
1395       1395        17        66         1         2         3        32   
1442       1442        12        26        11        47        10         9   
1500       1500        46        38         1        39       100        10   
1506       1506        19        26        24        37         5        28   
1695       1695        25       100        73         7         5         2   
1774       1774        46        43        54        82        12        40   
1802       1802        48        27       100        24        35         8   
1904       1904        32        18        75        66       100        13   
1920       1920        39        33         1        30        59         4   
1940       1940         3       100        80         5        94        13   
2344       2344        12        94        53       100        17         2   
2448       2448        27         5        23        60        25        80   
2463       2463        16        19       100         1        71        25   
2674       2674        18        34         1        19        32        47   
2675       2675        87       100        17        31        61        55   
2676       2676        27        87         1       100        73        25   
2701       2701        31         1        82        18        61        53   
2823       2823         1        15        60         4        33        39   
2915       2915        17        94        33         8       100        54   
3042       3042        31       100        65        66         1        68   
3161       3161         3       100         1        69        59        56   
3180       3180        26        47        19        82        61         1   
3181       3181        46        14        32        47        31        26   
3205       3205         1        41        45        81        17         3   
3251       3251        38        45         4        46         9        67   
3289       3289        12        29        17        19        75       100   
3395       3395         4        12       100         1        51        66   
3485       3485         1         6        95        19        46        14   
3718       3718        53       100         9        18        66        38   
3854       3854         7         1       100        18        60        22   
3868       3868         1        32        75        18        53         3   
3894       3894        27        26       100        25 

In [217]:
data.loc[4814].day=100
data.loc[1395].day=50
acct_cost(sced)

447.3233853883511

In [218]:
data.pref_rank.value_counts()

0    3556
1     454
2     214
3     183
5     128
4     108
7     100
8      92
6      88
9      77
Name: pref_rank, dtype: int64

In [219]:
data.assign(cost=family_cost(data.pref_rank,data.n_people))
data

TypeError: 'Series' objects are mutable, thus they cannot be hashed